// @smithy/middleware-content-length/dist-es/index.js
import '../../types/dist-es/auth/auth.js';
import '../../types/dist-es/auth/HttpApiKeyAuth.js';
import '../../types/dist-es/endpoint.js';
import '../../types/dist-es/extensions/checksum.js';
import '../../types/dist-es/http.js';
import '../../types/dist-es/profile.js';
import '../../types/dist-es/transfer.js';
import { HttpRequest } from '../../protocol-http/dist-es/httpRequest.js';

const CONTENT_LENGTH_HEADER = "content-length";
function contentLengthMiddleware(bodyLengthChecker) {
    return (next) => async (args) => {
        const request = args.request;
        if (HttpRequest.isInstance(request)) {
            const { body, headers } = request;
            if (body &&
                Object.keys(headers)
                    .map((str) => str.toLowerCase())
                    .indexOf(CONTENT_LENGTH_HEADER) === -1) {
                try {
                    const length = bodyLengthChecker(body);
                    request.headers = {
                        ...request.headers,
                        [CONTENT_LENGTH_HEADER]: String(length),
                    };
                }
                catch (error) {
                }
            }
        }
        return next({
            ...args,
            request,
        });
    };
}
const contentLengthMiddlewareOptions = {
    step: "build",
    tags: ["SET_CONTENT_LENGTH", "CONTENT_LENGTH"],
    name: "contentLengthMiddleware",
    override: true,
};
const getContentLengthPlugin = (options) => ({
    applyToStack: (clientStack) => {
        clientStack.add(contentLengthMiddleware(options.bodyLengthChecker), contentLengthMiddlewareOptions);
    },
});

export { contentLengthMiddleware, contentLengthMiddlewareOptions, getContentLengthPlugin };
