// @aws-sdk/middleware-host-header/dist-es/index.js
import '../../../@smithy/types/dist-es/auth/auth.js';
import '../../../@smithy/types/dist-es/auth/HttpApiKeyAuth.js';
import '../../../@smithy/types/dist-es/endpoint.js';
import '../../../@smithy/types/dist-es/extensions/checksum.js';
import '../../../@smithy/types/dist-es/http.js';
import '../../../@smithy/types/dist-es/profile.js';
import '../../../@smithy/types/dist-es/transfer.js';
import { HttpRequest } from '../../../@smithy/protocol-http/dist-es/httpRequest.js';

function resolveHostHeaderConfig(input) {
    return input;
}
const hostHeaderMiddleware = (options) => (next) => async (args) => {
    if (!HttpRequest.isInstance(args.request))
        return next(args);
    const { request } = args;
    const { handlerProtocol = "" } = options.requestHandler.metadata || {};
    if (handlerProtocol.indexOf("h2") >= 0 && !request.headers[":authority"]) {
        delete request.headers["host"];
        request.headers[":authority"] = request.hostname + (request.port ? ":" + request.port : "");
    }
    else if (!request.headers["host"]) {
        let host = request.hostname;
        if (request.port != null)
            host += `:${request.port}`;
        request.headers["host"] = host;
    }
    return next(args);
};
const hostHeaderMiddlewareOptions = {
    name: "hostHeaderMiddleware",
    step: "build",
    priority: "low",
    tags: ["HOST"],
    override: true,
};
const getHostHeaderPlugin = (options) => ({
    applyToStack: (clientStack) => {
        clientStack.add(hostHeaderMiddleware(options), hostHeaderMiddlewareOptions);
    },
});

export { getHostHeaderPlugin, hostHeaderMiddleware, hostHeaderMiddlewareOptions, resolveHostHeaderConfig };
