// @aws-sdk/nested-clients/dist-es/submodules/sts/runtimeConfig.shared.js
import '../../../../../@smithy/types/dist-es/auth/auth.js';
import '../../../../../@smithy/types/dist-es/auth/HttpApiKeyAuth.js';
import '../../../../../@smithy/types/dist-es/endpoint.js';
import '../../../../../@smithy/types/dist-es/extensions/checksum.js';
import '../../../../../@smithy/types/dist-es/http.js';
import '../../../../../@smithy/types/dist-es/profile.js';
import '../../../../../@smithy/types/dist-es/transfer.js';
import { NoOpLogger } from '../../../../../@smithy/smithy-client/dist-es/NoOpLogger.js';
import { parseUrl } from '../../../../../@smithy/url-parser/dist-es/index.js';
import { fromBase64 } from '../../../../../@smithy/util-base64/dist-es/fromBase64.js';
import { toBase64 } from '../../../../../@smithy/util-base64/dist-es/toBase64.js';
import { fromUtf8 } from '../../../../../@smithy/util-utf8/dist-es/fromUtf8.js';
import { toUtf8 } from '../../../../../@smithy/util-utf8/dist-es/toUtf8.js';
import { defaultSTSHttpAuthSchemeProvider } from './auth/httpAuthSchemeProvider.js';
import { defaultEndpointResolver } from './endpoint/endpointResolver.js';
import { AwsSdkSigV4Signer } from '../../../../core/dist-es/submodules/httpAuthSchemes/aws_sdk/AwsSdkSigV4Signer.js';
import { NoAuthSigner } from '../../../../../@smithy/core/dist-es/util-identity-and-auth/httpAuthSchemes/noAuth.js';

const getRuntimeConfig = (config) => {
    return {
        apiVersion: "2011-06-15",
        base64Decoder: config?.base64Decoder ?? fromBase64,
        base64Encoder: config?.base64Encoder ?? toBase64,
        disableHostPrefix: config?.disableHostPrefix ?? false,
        endpointProvider: config?.endpointProvider ?? defaultEndpointResolver,
        extensions: config?.extensions ?? [],
        httpAuthSchemeProvider: config?.httpAuthSchemeProvider ?? defaultSTSHttpAuthSchemeProvider,
        httpAuthSchemes: config?.httpAuthSchemes ?? [
            {
                schemeId: "aws.auth#sigv4",
                identityProvider: (ipc) => ipc.getIdentityProvider("aws.auth#sigv4"),
                signer: new AwsSdkSigV4Signer(),
            },
            {
                schemeId: "smithy.api#noAuth",
                identityProvider: (ipc) => ipc.getIdentityProvider("smithy.api#noAuth") || (async () => ({})),
                signer: new NoAuthSigner(),
            },
        ],
        logger: config?.logger ?? new NoOpLogger(),
        serviceId: config?.serviceId ?? "STS",
        urlParser: config?.urlParser ?? parseUrl,
        utf8Decoder: config?.utf8Decoder ?? fromUtf8,
        utf8Encoder: config?.utf8Encoder ?? toUtf8,
    };
};

export { getRuntimeConfig };
